File: leetcode/python/000347.top-k-frequent-elements.py
class Solution:
    # Time complexity: O(n log n)
    # Space complexity: O(n)
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        map: dict[int, int] = {}
        for n in nums:
            map[n] = map.get(n, 0) + 1

        items = sorted(map.items(), key=lambda t: t[1], reverse=True)

        res = []
        for i in range(k):
            res.append(items[i][0])

        return res


class Solution2:
    # Using a variation of bucket sort
    # Time complexity: O(n)
    # Space complexity: O(n)
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        count: dict[int, int] = {}
        freq: list[list[int]] = [[] for _ in range(len(nums) + 1)]

        for n in nums:
            count[n] = count.get(n, 0) + 1
        for n, c in count.items():
            freq[c].append(n)

        res: list[int] = []
        for i in range(len(freq) - 1, 0, -1):
            for n in freq[i]:
                res.append(n)
                if len(res) == k:
                    return res


## IN PROGRESS

File: leetcode/javascript/000347.top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 * Time complexity: O(n)
 * Space complexity: O(n)
 */
var topKFrequent = function (nums, k) {
  const count = new Map();
  const freq = Array.from({ length: nums.length + 1 }, () => []);

  for (const num of nums) {
    count.set(num, (count.get(num) ?? 0) + 1);
  }

  count.forEach((val, key) => {
    freq[val].push(key);
  });

  const res = [];
  for (let i = freq.length - 1; i >= 0; i--) {
    for (const n of freq[i]) {
      res.push(n);
      if (res.length === k) {
        return res;
      }
    }
  }
};
